lms/djangoapps/courseware/tabs.py

```diff
@@ -3,44 +3,346 @@
+            """ Returns the link_value as the link. """
+            return self.link_value
+
+        self.type = tab_dict['type']
+
+        super(LinkTab, self).__init__(
+            name=tab_dict['name'] if tab_dict else name,
+            tab_id=None,
+            link_func=link_value_func,
+        )
+
+    def __getitem__(self, key):
+        if key == 'link':
+            return self.link_value
+        else:
+            return super(LinkTab, self).__getitem__(key)
+
+    def __setitem__(self, key, value):
+        if key == 'link':
+            self.link_value = value
+        else:
+            super(LinkTab, self).__setitem__(key, value)
+
+    def to_json(self):
+        to_json_val = super(LinkTab, self).to_json()
+        to_json_val.update({'link': self.link_value})
+        return to_json_val
+
+    def __eq__(self, other):
+        if not super(LinkTab, self).__eq__(other):
+            return False
+        return self.link_value == other.get('link')
+
+
+class SingleTextbookTab(CourseTab):
+    """
+    A tab representing a single textbook.  It is created temporarily when enumerating all textbooks within a
+    Textbook collection tab.  It should not be serialized or persisted.
+    """
+    type = 'single_textbook'
+    is_movable = False
+    is_collection_item = True
+    priority = None
+
+    def to_json(self):
+        raise NotImplementedError('SingleTextbookTab should not be serialized.')
+
+
+def get_course_tab_list(request, course):
     """
     Retrieves the course tab list from xmodule.tabs and manipulates the set as necessary
     """
-    user_is_enrolled = user.is_authenticated() and CourseEnrollment.is_enrolled(user, course.id)
-    xmodule_tab_list = CourseTabList.iterate_displayable(
-        course,
-        settings,
-        user.is_authenticated(),
-        has_access(user, 'staff', course, course.id),
-        user_is_enrolled
-    )
-
-    # Now that we've loaded the tabs for this course, perform the Entrance Exam work
-    # If the user has to take an entrance exam, we'll need to hide away all of the tabs
-    # except for the Courseware and Instructor tabs (latter is only viewed if applicable)
-    # We don't have access to the true request object in this context, but we can use a mock
-    request = RequestFactory().request()
-    request.user = user
+    user = request.user
+    xmodule_tab_list = CourseTabList.iterate_displayable(course, user=user)
+
+    # Now that we've loaded the tabs for this course, perform the Entrance Exam work.
+    # If the user has to take an entrance exam, we'll need to hide away all but the
+    # "Courseware" tab. The tab is then renamed as "Entrance Exam".
     course_tab_list = []
     for tab in xmodule_tab_list:
         if user_must_complete_entrance_exam(request, user, course):
-            # Hide all of the tabs except for 'Courseware' and 'Instructor'
+            # Hide all of the tabs except for 'Courseware'
```

Comment:
Interesting note: all dynamic tabs will now be shown for a course with an unfinished entrance exam. I think this is okay, because such tabs are role-based, so today that's just the "CCX" and "Instructor" tabs.

I wonder whether the entrance exam hiding of tabs could now be implemented in ``is_enabled``. Maybe ``EnrolledCourseViewType`` could return ``False`` if a user hasn't passed an entrance exam yet. That would be cleaner than having this magic hack to remove the tabs. Unfortunately the code would still need the magic renaming hack, unless we also provided a way to have dynamic tab names.